components/eventing-controller/controllers/subscription/jetstream/reconciler_test.go

```diff
@@ -304,6 +335,131 @@ func TestChangeSubscription(t *testing.T) {
 		changeSubscription    func(subscription *eventingv1alpha1.Subscription)
 		wantAfter             utils.Want
 	}{
+		{
+			// Ensure subscriptions on NATS are added when adding a Subscription filter and providing an invalid sink.
```

Comment:
To understand it a bit easier I wrote a small test to make the problem more obvious:

```go
{
	name: "Test if NATS subscription is created then having a valid event-type but invalid sink",
	givenSubscriptionOpts: []reconcilertesting.SubscriptionOpt{
		reconcilertesting.WithFilter(reconcilertesting.EventSource, utils.NewCleanEventType("0")),
		reconcilertesting.WithSinkURL("invalid"),
	},
	want: utils.Want{
		K8sSubscription: []gomegatypes.GomegaMatcher{
			reconcilertesting.HaveSubscriptionNotReady(),
			reconcilertesting.HaveCleanEventTypes([]string{
				utils.NewCleanEventType("0"),
			}),
		},
		NatsSubscriptions: map[string][]gomegatypes.GomegaMatcher{
			utils.NewCleanEventType("0"): {
				natstesting.BeExistingSubscription(),
				natstesting.BeValidSubscription(),
				natstesting.BeJetStreamSubscriptionWithSubject(utils.NewCleanEventType("0")),
			},
		},
	},
},
```

The test is green with the current code but was red before. So THIS is a change in the BEHAVIOUR:

<details>
  <summary>☑️ With current code</summary>

```
=== RUN   TestCreateSubscription/Test_if_NATS_subscription_is_created_when_having_a_valid_event-type_but_invalid_sink
    reconciler_test.go:321: creating the k8s subscription
    reconciler_test.go:323: testing the k8s subscription
{"level":"ERROR","timestamp":"2022-07-06T14:42:58+02:00","logger":"jetstream-subscription-reconciler","caller":"jetstream/reconciler.go:156","message":"Failed to validate sink URL","context":{"kind":"Subscription","version":1,"namespace":"test","name":"nats-sub-0","error":"sink URL scheme should be 'http' or 'https'"}}
{"level":"ERROR","timestamp":"2022-07-06T14:42:58+02:00","logger":"jetstream-subscription-reconciler","caller":"jetstream/reconciler.go:156","message":"Failed to validate sink URL","context":{"kind":"Subscription","version":1,"namespace":"test","name":"nats-sub-0","error":"sink URL scheme should be 'http' or 'https'"}}
    reconciler_test.go:326: testing the k8s events
    reconciler_test.go:329: testing the nats subscriptions
    reconciler_test.go:334: testing the deletion of the subscription
    reconciler_test.go:337: testing the deletion of the NATS subscription(s)
```

</details> 

<details>
  <summary>🔴 With code from main</summary>
=== RUN   TestCreateSubscription/Test_if_NATS_subscription_is_created_when_having_a_valid_event-type_but_invalid_sink
    reconciler_test.go:321: creating the k8s subscription
    reconciler_test.go:323: testing the k8s subscription
{"level":"ERROR","timestamp":"2022-07-06T14:45:43+02:00","logger":"jetstream-subscription-reconciler","caller":"jetstream/reconciler.go:147","message":"sink URL validation failed","context":{"kind":"Subscription","version":1,"namespace":"test","name":"nats-sub-0","error":"sink URL scheme should be 'http' or 'https'"}}
{"level":"ERROR","timestamp":"2022-07-06T14:45:43+02:00","logger":"jetstream-subscription-reconciler","caller":"jetstream/reconciler.go:147","message":"sink URL validation failed","context":{"kind":"Subscription","version":1,"namespace":"test","name":"nats-sub-0","error":"sink URL scheme should be 'http' or 'https'"}}
    reconciler_test.go:326: testing the k8s events
    reconciler_test.go:329: testing the nats subscriptions
{"level":"ERROR","timestamp":"2022-07-06T14:45:47+02:00","logger":"jetstream-subscription-reconciler","caller":"jetstream/reconciler.go:147","message":"sink URL validation failed","context":{"kind":"Subscription","version":1,"namespace":"test","name":"nats-sub-0","error":"sink URL scheme should be 'http' or 'https'"}}
{"level":"ERROR","timestamp":"2022-07-06T14:45:49+02:00","logger":"jetstream-subscription-reconciler","caller":"jetstream/reconciler.go:147","message":"sink URL validation failed","context":{"kind":"Subscription","version":1,"namespace":"test","name":"nats-sub-0","error":"sink URL scheme should be 'http' or 'https'"}}
{"level":"ERROR","timestamp":"2022-07-06T14:45:51+02:00","logger":"jetstream-subscription-reconciler","caller":"jetstream/reconciler.go:147","message":"sink URL validation failed","context":{"kind":"Subscription","version":1,"namespace":"test","name":"nats-sub-0","error":"sink URL scheme should be 'http' or 'https'"}}
{"level":"ERROR","timestamp":"2022-07-06T14:45:53+02:00","logger":"jetstream-subscription-reconciler","caller":"jetstream/reconciler.go:147","message":"sink URL validation failed","context":{"kind":"Subscription","version":1,"namespace":"test","name":"nats-sub-0","error":"sink URL scheme should be 'http' or 'https'"}}
    reconciler_test.go:776: 
        Timed out after 10.000s.
        nats subscriptions should be as defined
        Expected
            <bool>: false
        to be true
```

</details>

Reply:
Let's assume we would not create the subscription on NATS side and see how this would change the newly added test:


- `NATS subscriptions are in sync even though the sink becomes invalid: adding a filter`: The test is similar to the test above except that this test already has a NATS subscription before the sink is rendered invalid in the `changeSubscription` test case entry. If we would not create the NATS subscription in the simple case above, we should also not do it in the mentioned test case.
- `NATS subscriptions are in sync even though the sink becomes invalid: deleting a filter`: if we don't want to have NATS subscriptions where we have an invalid sink, then we should delete the NATS subscription, so the test would stay unchanged.